[GH-862] Added issue status field in subscription modal

raghavaggarwal2308 opened the pull request:
#### Summary
Added issue status field in the subscription modal and created an API to fetch the possible statuses of a project.

#### Screenshots
![image](https://github.com/Brightscout/mattermost-plugin-jira/assets/55234496/df54701d-bbea-4645-9627-e4223f600989)

#### What to test?
*Note:* The below things should be tested for both cloud and server instances of Jira.
- The status field is properly displayed in the subscription modal.
- The status field can be updated and are being sent in the subscription payload.
- The status field is properly prefilled while editing a subscription.
- The status field is only visible when a project is selected.
- The options in the status fields are filtered properly according to project and issue type.
- Webhook notifications are received according to the subscription filters.

###### Steps to reproduce
1. Connect your Jira account.
2. Run the `/jira subscribe` command and create/edit a subscription.
3. Save/create the subscription.
4. Generate a notification from the Jira side according to the subscription.

###### Environment:
MM version: v7.8.2
Node version: 14.18.0
Go version: 1.19.0

#### Ticket Link
Fixes #862

mickmister commented:
@raghavaggarwal2308 I'm thinking this should go in the field filters part of the modal, rather than have its own place under the issue type selector. This way you can selectively include/exclude certain statuses. What are the pros and cons of this approach versus the current one?

raghavaggarwal2308 commented:
> @raghavaggarwal2308 I'm thinking this should go in the field filters part of the modal, rather than have its own place under the issue type selector. This way you can selectively include/exclude certain statuses. What are the pros and cons of this approach versus the current one?

@mickmister There are four options in the filters section i.e. include, exclude, include all, and empty. If we add the status field in the filters section then the "include all" and "empty" options will be useless as I think the status field can't have multiple values at a time and it can't be empty as well.
We have three solutions for this:
1. We move the field to the filters section and it will be the user's responsibility to be careful while using the two options for "Status filter".
2. We conditionally remove the two options from the dropdown.
3. We leave the field as it is now (i.e. below the issue type selector) and the user can select the statuses he wants to get notified for.

The advantage we will have by moving the field to the filters section is that the user will be able to use the "exclude" option.

raghavaggarwal2308 commented:
> The field "filters" are meant for "what state is the ticket currently in", which the status certainly applies to that. Thoughts on moving it to the filters?

@mickmister Sounds good, will make the change and update the PR.

raghavaggarwal2308 commented:
@mickmister Updated:
- Moved the issue status field into the filters section of the modal.
- Conditionally removed the "Empty" and "Include all" options.

Screenshot:
![image](https://github.com/Brightscout/mattermost-plugin-jira/assets/55234496/df54701d-bbea-4645-9627-e4223f600989)

mickmister commented on server/client_cloud.go:
```diff
@@ -132,3 +133,17 @@ func (client jiraCloudClient) GetIssueTypes(projectID string) ([]jira.IssueType,
 	return result, nil
 }
+
+type IssueType struct {
```
```suggestion
type IssueTypeWithStatuses struct {
```

mickmister commented on webapp/src/components/modals/channel_subscriptions/edit_channel_subscription.tsx:
```diff
@@ -238,6 +268,40 @@ export default class EditChannelSubscription extends PureComponent<Props, State>
     };
 
+    fetchIssueStatuses = (projectID: string, instanceID: string) => {
+        if (!instanceID) {
+            this.setState({getIssueStatusesErr: 'No Jira instance is selected.'});
+            return;
+        }
```
Under what circumstance would this happen, if ever?

mickmister commented on webapp/src/components/modals/channel_subscriptions/edit_channel_subscription.tsx:
```diff
@@ -211,7 +219,29 @@ export default class EditChannelSubscription extends PureComponent<Props, State>
+        const keys: string[] = [];
+        const statuses: Status[] = [];
+        if (this.state.projectStatuses) {
+            this.state.projectStatuses.forEach((element: ProjectStatuses) => {
```
This variable name and type name are confusing me a bit. Is this an accurate understaning?

```suggestion
            this.state.projectStatuses.forEach((issueType: IssueTypeWithStatuses) => {
```

mickmister commented on server/subscribe.go:
```diff
@@ -148,6 +149,10 @@ func (p *Plugin) matchesSubsciptionFilters(wh *webhook, filters SubscriptionFilt
 		return false
 	}
 
+	if filters.IssueStatuses.Len() != 0 && !filters.IssueStatuses.ContainsAny(issue.Fields.Status.ID) {
+		return false
+	}
```
We also need to check for exclude right?

Also, are we able to put this into the `SubscriptionFilters.Fields` instead of making a separate `SubscriptionFilters.IssueStatuses`? This way we can reuse some logic for comparing field values below

mickmister commented on webapp/src/components/modals/channel_subscriptions/edit_channel_subscription.tsx:
```diff
@@ -66,10 +66,13 @@ export type State = {
     subscriptionName: string | null;
     showConfirmModal: boolean;
     conflictingError: string | null;
+    issue_statuses: Status[] | null;
+    projectStatuses: ProjectStatuses[] | null;
```
The mix of snake case and camel case seems weird here. All of the other keys are camel case so I'm thinking we should go with that

mickmister commented on webapp/src/components/modals/channel_subscriptions/edit_channel_subscription.tsx:
```diff
@@ -275,9 +340,14 @@ export default class EditChannelSubscription extends PureComponent<Props, State>
         }
 
+        if (projects && projects.length) {
+            this.fetchIssueStatuses(projects[0], this.state.instanceID);
```
We can put this in the same if block as above

Also, we could include the statuses in the response to `fetchIssueMetadata` instead of creating a new endpoint right? The statuses are currently only fetched at the same spot as that call anyway. That may simplify code a bit and require less code for this feature

mickmister replied:
Another approach would be to encapsulate the fetching and state management to a separate component. This way the parent component doesn't need to change at all. This is currently how the epic link selector works by delegating the state management to the `BackendSelector` component.

https://github.com/mattermost/mattermost-plugin-jira/blob/d1e233af901f5a4c3b652fee260857c74d1c6d44/webapp/src/components/data_selectors/jira_epic_selector/jira_epic_selector.tsx#L100-L106

mickmister reviewed (commented):
Awesome work on this @raghavaggarwal2308 :+1: I have a few suggestions, one being that we include the status information in the `create-issue-metadata` call to cut down on the frontend code needed to fetch the statuses separately. Please let me know your thoughts thanks

raghavaggarwal2308 commented:
@mickmister Fixed the review comments mentioned by you. We are now returning the status data from the create meta data API.

mickmister commented:
@raghavaggarwal2308 Can you please update the screenshot in the PR description to match the current code?

mickmister commented on server/issue.go:
```diff
@@ -375,16 +380,29 @@ func (p *Plugin) httpGetCreateIssueMetadataForProjects(w http.ResponseWriter, r
 	}
 
-	return client.GetCreateMetaInfo(p.API, &jira.GetQueryOptions{
+	projectStatuses, err := client.ListProjectStatuses(projectKeys)
```
Have we tested this feature on Jira Cloud and Jira Server?

raghavaggarwal2308 replied:
@mickmister Tested and made the necessary changes

mickmister commented on webapp/src/types/model.ts:
```diff
@@ -65,16 +65,17 @@ export type Project = {
+export type IssueMetadata = {
+    projects: Project[];
+    issue_types: IssueTypeWithStatuses[];
+}
```
Is the sole purpose of having a separate `issue_types` here to have the statuses? Any way we can just have the backend update `project.issuetypes` to include the statuses? It seems like we may be returning twice the data in this code by duplicating the issue types. If we keep them separate then I think we should at least remove everything from `IssueTypeWithStatuses` that doesn't have to do with statuses before sending back to the client, if statuses is all it's being used for.

Edit: I think I understood this wrong. The `ListProjectStatuses` backend `JiraClient` function doesn't return all of the issue type data. I'll trust your judgment how this should be organized. I think we should be more specific with this name though like `issue_types_with_statuses`.

raghavaggarwal2308 replied:
@mickmister Yes, `ListProjectStatuses` returns only the fields that we need in our implementation. It would be very complex to move this inside the `project.issueTypes` as we have to manually iterate over both the array multiple times and add the field. Also, the type used for `issueTypes` inside the project data does not have the "statuses" field. So, we have to create a new custom type instead of using the type provided by the package.

mickmister commented on webapp/src/utils/jira_issue_metadata.tsx:
```diff
@@ -224,6 +228,44 @@ export function getCustomFieldFiltersForProjects(metadata: IssueMetadata | null,
         } as FilterField);
     }
 
+    // Filtering out the statuses on the basis of selected issue types
+    const issueTypesWithStatuses = metadata && metadata.issue_types;
+    const keys = new Set<string>();
+    const statuses: Status[] = [];
```
I think this and code below could be extracted into a separate function. It could return `statusField | undefined`, then this function could conditionally add it to `result` if it's not `undefined`.

We should also edit the existing tests in [jira_issue_metadata.test.tsx](https://github.com/mattermost/mattermost-plugin-jira/blob/master/webapp/src/utils/jira_issue_metadata.test.tsx) to include checks for statuses

---

I wrote the below comment before I noticed that there are indeed tests for this file:

It would be great to have `getCustomFieldFiltersForProjects` or others in this file unit tested but not necessary right now. This is the sort of task that potentially makes a good HW ticket. If you see an potential improvement to the code such as code missing unit tests, I encourage you to create a ticket to document your thought and put the "Help Wanted" etc. labels on the ticket. "Write tests for selectors in jira_issue_metadata.tsx`

mickmister reviewed (commented):
A few more comments for discussion

raghavaggarwal2308 commented:
@mickmister Fixed the review comments mentioned by you

raghavaggarwal2308 commented:
@hanzei Resolved conflicts

mickmister commented on webapp/src/utils/jira_issue_metadata.tsx:
```diff
@@ -190,6 +190,47 @@ function isValidFieldForFilter(field: JiraField): boolean {
+    const keys = new Set<string>();
+    const statuses: Status[] = [];
+    if (issueTypesWithStatuses) {
+        for (const issueType of issueTypesWithStatuses) {
+            if (issueTypes.includes(issueType.id) || !issueTypes.length) {
+                for (const status of issueType.statuses) {
+                    if (!keys.has(status.id)) {
+                        keys.add(status.id);
+                        statuses.push(status);
```
Will `issueTypes` ever have length 0? (looking at `!issueTypes.length`)

Also it's a bit hard to read will all of the variations of:
- issueTypes
- issueTypesWithStatuses
- issueType
- issueType.statuses

Not sure there is a better way though. I was going to suggest renaming the function parameter `issueTypes` to `issueTypesWithoutStatuses`, but that sounds like "these issue types have no statuses associated with them". Current code seems fine, just a lot of redundancy causing a bit of confusion for me.

raghavaggarwal2308 replied:
> Will issueTypes ever have length 0? (looking at !issueTypes.length)
@mickmister Yes, when we have not selected any value in the modal.

mickmister replied:
Maybe we can change the name to `selectedIssueTypes` to convey they are user set values

raghavaggarwal2308 replied:
> I will rename issueTypes to selectedIssueTypes for more clarity.
@mickmister Already changed

mickmister commented on webapp/src/utils/jira_issue_metadata.test.tsx:
```diff
@@ -293,6 +293,105 @@ describe('utils/jira_issue_metadata', () => {
 
+    test('getStatusField should return options for statuses for selected issue types only', () => {
+        const metadata: IssueMetadata = {
+            projects: [
+                {
+                    key: 'TEST',
+                    issuetypes: [],
+                },
+            ],
```
Is it realistic to have a blank array for `issuetypes` here?

raghavaggarwal2308 replied:
@mickmister No, I don't think this will ever be blank. This particular field is not useful in the function we are testing here, so we kept it empty. I you want we can add some mock data or maybe just add a comment above it.

mickmister replied:
We should try to have realistic test data I suppose. I recommend using the saved test data when possible

raghavaggarwal2308 replied:
@mickmister Updated the test files

mickmister commented on webapp/src/utils/jira_issue_metadata.test.tsx:
```diff
@@ -293,6 +293,105 @@ describe('utils/jira_issue_metadata', () => {
+        }
+    });
+
+    test('getStatusField should return options for statuses for all issue types if no issue type is selected', () => {
```
What does "no issue type is selected" mean here?

mickmister reviewed (approved):
LGTM, just a few comments for discussion

raghavaggarwal2308 commented on server/client_server.go:
```diff
@@ -211,7 +211,7 @@ func (client jiraServerClient) GetIssueTypes(projectID string) ([]jira.IssueType
 func (client jiraServerClient) ListProjectStatuses(projectID string) ([]*IssueTypeWithStatuses, error) {
 	var result []*IssueTypeWithStatuses
-	if err := client.RESTGet(fmt.Sprintf("3/project/%s/statuses", projectID), nil, &result); err != nil {
+	if err := client.RESTGet(fmt.Sprintf("2/project/%s/statuses", projectID), nil, &result); err != nil {
```
@mickmister It was a mistake on my side, the latest API version for the Jira server is `2`.

mickmister replied:
Okay we definitely want to test any changes made to Jira Server code, with Jira Server. And mention in the test steps of the PR to test both Jira Cloud/Server depending on the changes

arush-vashishtha commented:
Tested the PR and is working fine for the following events,

- Issue updated
- Issue deleted
- Comment created
- Comment updated
- Comment deleted

Nonetheless in the event `issue created` when creating a subscription with the filter applied for `status`, the notifications are being generated for all the status in the subscribed channel or DM/GM on MM .
cc @raghavaggarwal2308

raghavaggarwal2308 commented:
@arush-vashishtha The issue is fixed now.

@mickmister The issue was caused because of an invalid payload being sent from the Jira side in case of issue created event. The status field was always sending the data of a specific status ("Backlog" in our case) irrespective of the status selected on Jira. I have updated the logic to fetch the proper data of the issue in case of this event. Also added a TODO to update the logic in case we get proper data from Jira in the future.

AayushChaudhary0001 commented:
The reported issue is fixed and tested now, working fine for this one and all the events which are already mentioned before. Approved.

mickmister commented:
HEads up that there are conflicts to resolve here

mickmister commented:
@arush-vashishtha Were you able to test this with both Jira Cloud and Jira Server?

arush-vashishtha commented:
> @arush-vashishtha Were you able to test this with both Jira Cloud and Jira Server?

Yes I have tested this on both Jira Cloud and Jira Server.
